### DISCUSSION TITLE
Normal map provided with cube model  is in a tangent space of OpenGL style?

### DISCUSSION THREAD

**Discussion by wojtekpil** [sotrh/learn-wgpu#324] [Q&A] [answered]
It seems that provided normal map textures for tutorial 11 are using tangent space of OpenGL style. WGPU uses coordinate system of D3D & Metal so results seems to be incorrect. I tested it with white diffuse texture:
![Screenshot from 2022-02-23 20-14-46](https://user-images.githubusercontent.com/1286923/155393188-1311bfd3-8c07-47c6-93f9-9696e1c805ee.png)
And here is a result with Y channel inverted of normalmap (D3D normalmap): 
![Screenshot from 2022-02-23 20-15-24](https://user-images.githubusercontent.com/1286923/155393312-3a533ed7-5eb3-4925-8fb7-b329cd9c16dc.png)

Its less noticeable with diffuse texture.
Original (opengl):
![Screenshot from 2022-02-23 20-02-50](https://user-images.githubusercontent.com/1286923/155393515-9b653b53-b888-40ea-b654-046bb3066e2c.png)
D3D converted:
![Screenshot from 2022-02-23 20-00-52](https://user-images.githubusercontent.com/1286923/155393573-7e3f512d-719b-4947-8754-25c9ea37fd11.png)

I couldn't test it with provided demo source code (some weird crash I need to investigate), so it could be mistake in my code, but the issue is similar to what we can see in images from tutorial.

**Comment by sotrh** [comment, marked as answer]
You are correct. Another user pointed this out in #313. Do you know of any tools that can convert the cube normal textures to the correct handedness?

**Comment by wojtekpil** [comment]
Oh, you are right, somehow I didn't notice this issue. Usually I am using open-source material-maker. You can use this node setup:
![Screenshot from 2022-02-26 21-49-53](https://user-images.githubusercontent.com/1286923/155858682-ed25fe73-f254-41e7-a0c0-3e2718316f66.png)
Probably there are easier methods, but I am using material-maker for procedural textures so I am used to it.
